Lightning Web Components (#192)
coveo/source.coveo.com: _posts/2019-04-18-lightning-web-components.md

---
layout: post

title: "Coveo and the Lightning Web Components framework"
tags: [Salesforce, Lightning Web Components, LWC, Framework]

author:
  name: Etienne Rocheleau & Andre Theriault & Jeremie Robert & Louis Bompart
  bio: Software Developers
  twitter: 
  image: erocheleau.jpg

---

In 2018, Salesforce introduced a new programming model for building Lightning components. As you may already know, Lightning components are the building blocks used to create applications within Salesforce. This new framework was designed with respect to the new web standards in order to allow web developers to keep using the knowledge they have acquired over the years outside of Salesforce. This new building model is alligned with the Web Components standard from the World Wide Web Consortium (W3C) and is using native browser features and language such as pure HTML and JavaScript. This blog post will document our team's experience with this new framework to build the Coveo for Salesforce integration using Lightning Web Components (LWC).

> <!-- more -->

Reviewer: I think you should do a bit more to explain that this post is written as an interview of the developers. Right now you hint at this only in the last sentence of a big paragraph, and even then it's not so obvious. I know that I was a bit surprised by the format afterwards when reading.

## What does this Lightning Web Components name mean?

> Lightning Web Components include two concepts in its name.

Reviewer: ```suggestion
Lightning Web Components include two concepts in its name:
```

**Lightning**: It is a Salesforce related term which defines their UI or front-end applicative framework (by opposition to "Classic" or "Visualforce"). 

**Web Components**: These are the result of three technologies that, when used in combination, allow you to define custom and reusable web elements. It leverages *Custom elements* which allows you to define custom web elements in JavaScript and some behaviour which in turn can be rendered in a web browser. These custom elements also use the *Shadow DOM* which is an encapsulation of the DOM tree attached to a specific element. This is used to isolate features and behaviour of your element from other parts of the document. And finally, Web Components use the `<template>` and `<slot>` elements that allow you to write markup that is not displayed in your page but can be reused where you want in your page.

So if you put the two together, you get Web Components that can interact with Salesforce, either to access data or to be part of their application rendering loop. Which means Salesforce has made available this new bleeding edge web technology (Web Components) and allowed it to communicate very easily with Salesforce (Lightning).

> ## Meet the team

Reviewer: What no pictures? :)

### Andre Theriault

> I've been part of the Salesforce Integration team for 4 years now as a developer. Most of my experience before that was as a back-end software engineer. With Salesforce's efforts to use newer techs and up to date web standards, I learned to enjoy the front-end aspect of it all. It also got me interested in other frontend frameworks like React and Vue.js and I was pleased to discover that Salesforce was planning to get more inline with current web standards with LWC.

### Etienne Rocheleau

> At Coveo for 4 years, and my job consists of the exploration of all new technology from Salesforce. I am always looking for new ways to integrate Coveo with Salesforce to create a better experience for users. I produce a lot of demos, POCs, and prototypes as part of my exploration and I really love when people find new and interesting ways to use Coveo to solve an issue they have. I am very familiar with Salesforce technologies (Aura, Apex, Visualforce, Lightning, etc.) and outside of that, I have experience with JavaScript/TypeScript and some web frameworks such as React, Angular, and others.

> ## Why switch to LWC when you are so involved with Aura?

Reviewer: The CSS doesn't make a big difference between the heading levels - it makes the structure a bit harder to see IMO.

Author: Yea resolved that with a css proposed change. All quotes were in **bold** by default, removed that styling. It looks better now.

\- Louis

> Some gotchas to take into account that had me searching for a little while:
>
> 1. As of writing this post, LWC does not allow for custom decorators, only the ones provided by Salesforce and the LWC framework. JavaScript decorators are still only a proposition to ECMAScript, but seeing the decorators offered by Salesforce makes me want to be able to create my own. For example, I would like to create a `@log({level: 'info'})` decorator to mark a method to log to the console for tracking purposes. Or one of the frequently seen use cases with LWC: `@executeOnce` could be useful too.

> > 2. Be wary of the difference between `this.querySelector`, `this.template.querySelector`, and `element.shadowRoot.querySelector`. The first one is used to search the DOM in a *slot*, the second one is used to search within the body of your LWC, and the third one is used within unit tests.

Reviewer 2: `element.shadowRoot.querySelector` is equivalent to `this.template.querySelector` and `element.querySelector` is equivalent to `this.querySelector`. The first queries the shadow DOM and the later queries the light DOM. It's probably something we should enhance in our documentation.

Maybe this section should be reformulated not to confuse readers.

Author: Okay this was based on our experience and something we saw in documentation around unit tests with Jest.
You are right, after looking at examples, we were only using `element.shadowRoom.querySelector` because we are creating ourselves the element with 
`const element = createElement();`
Thank you, I will probably remove that part since `element.` is confusing indeed.
I still like the difference between the `this.querySelector` and `this.template.querySelector` though, what do you think?
It comes from here:
https://developer.salesforce.com/docs/component-library/documentation/lwc/create_components_slots

---
layout: post

title: "Coveo and the Lightning Web Components framework"
tags: [Salesforce, Lightning Web Components, LWC, Framework]

author:
  name: Etienne Rocheleau & Andre Theriault & Jeremie Robert & Louis Bompart
  bio: Software Developers
  twitter: 
  image: erocheleau.jpg

---

> In 2018, Salesforce introduced a new programming model for building Lightning components (Lightning components are the building blocks used to create applications within Salesforce). This new framework was designed with respect to the new web standards in order to allow web developers to keep using the knowledge they have acquired over the years outside of Salesforce. This new building model is alligned with the Web Components standard from the World Wide Web Consortium (W3C) and is using native browser features and language such as pure HTML and JavaScript. This blog post will document our team's experience with this new framework to build the Coveo for Salesforce integration using Lightning Web Components (LWC).

Reviewer: ```suggestion
In 2018, Salesforce introduced a new programming model for building Lightning components. As you might know already, Lightning components are the building blocks used to create applications within Salesforce. This new framework was designed with respect to the new web standards in order to allow web developers to keep using the knowledge they have acquired over the years outside of Salesforce. This new building model is aligned with the Web Components standard from the World Wide Web Consortium (W3C) and is using native browser features and language such as pure HTML and JavaScript. This blog post will document our team's experience with this new framework to build the Coveo for Salesforce integration using Lightning Web Components (LWC).
```

<!-- more -->

> ## What is this Lightning Web Component thing?

Reviewer: ```suggestion
## What is this Lightning Web Components thing?
```

> Lightning Web Component includes two concepts in its name.

Reviewer: ```suggestion
Lightning Web Components includes two concepts in its name.
```

> The **Lightning** part of the *Lightning Web Components*. Lightning is a Salesforce related term which defines their UI or front-end applicative framework (by opposition to "Classic" or "Visualforce"). 

Reviewer: ```suggestion
The **Lightning** part is a Salesforce related term which defines their UI or front-end applicative framework (by opposition to "Classic" or "Visualforce"). 
```

> Second, **Web Components** are the result of three technologies that, when used in combination, allow you to define custom and reusable web elements. It leverages *Custom elements* which allows you to define custom web elements in JavaScript and some behaviour which in turn can be rendered in a web browser. These custom elements also use the *Shadow DOM* which is an encapsulation of the DOM tree attached to a specific element. This is used to separate features and behaviour of your element from other parts of the document. And finally, Web Components use the `<template>` and `<slot>` elements that allow you to write markup that is not displayed in your page but can be reused where you want in your page.

Reviewer: Separate or isolate?

Author: the correct term is probably encapsulate :)

Author: But yea I like isolate better here.

> Second, **Web Components** are the result of three technologies that, when used in combination, allow you to define custom and reusable web elements. It leverages *Custom elements* which allows you to define custom web elements in JavaScript and some behaviour which in turn can be rendered in a web browser. These custom elements also use the *Shadow DOM* which is an encapsulation of the DOM tree attached to a specific element. This is used to separate features and behaviour of your element from other parts of the document. And finally, Web Components use the `<template>` and `<slot>` elements that allow you to write markup that is not displayed in your page but can be reused where you want in your page.

Author: I was trying not to have a formal 1. 2. 3. basically

IMO there are much better sources of info on what a Web component is and I am trying to keep it at bare minimum here, see the first link at the bottom of the post.

So if you put the two together, you get Web Components that can interact with Salesforce, either to access data or to be part of their application rendering loop. Which means Salesforce has made available this new bleeding edge web technology (Web Components) and allowed it to communicate very easily with Salesforce (Lightning).

## Meet the team

Before jumping on our experience with LWC, here are the members of our team that took part in the original exploration of this new framework.

> ### Jeremie Robert

Reviewer: ```suggestion
### Jérémie Robert
```

Author: Yea... I removed accent's from everyone's names. Do you think I should keep them?

André
Jérémie
Étienne

:)

Reviewer: As you wish I guess. Still triggers me :)

> > At Coveo for over a year, I am constantly working at better integrating Coveo with the Salesforce platform. Using Salesforce technologies (Apex, Visualforce, Aura, and now LWC) and other web technologies (ES6, TypeScript, CSS, and SASS) I build solutions that can be personalized by our clients and that respect the highest standards in terms of quality.

Reviewer: no less 😛

Author: Nothing but the best

> > At Coveo for over a year, I am constantly working at better integrating Coveo with the Salesforce platform. Using Salesforce technologies (Apex, Visualforce, Aura, and now LWC) and other web technologies (ES6, TypeScript, CSS, and SASS) I build solutions that can be personalized by our clients and that respect the highest standards in terms of quality.

Author: Maybe, I thought about the pictures, didn't want to make it too long.

> > At Coveo for over a year, I am constantly working at better integrating Coveo with the Salesforce platform. Using Salesforce technologies (Apex, Visualforce, Aura, and now LWC) and other web technologies (ES6, TypeScript, CSS, and SASS) I build solutions that can be personalized by our clients and that respect the highest standards in terms of quality.

Author: I really wanted the quotes since it was the own words from every team member.
Any other idea to keep quotes but no bold? indeed it makes it look a lot better in regards to style if I remove the bold.

### Louis Bompart

> I've been a Software Developer at Coveo since August 2017, working mostly on the front-end side of the Salesforce integration and also the tooling associated with that project. I've started my web development learning process with [Polymer](https://www.polymer-project.org/) around early 2015 and I've been convinced of the potential of Web Components since then.

### Andre Theriault

> I've been part of the Salesforce Integration team for 4 years now as a developer. Most of my experience before that was as a backend software engineer. With Salesforce's efforts to use newer techs and up to date web standards, I learned to enjoy the frontend aspect of it all. It also got me interested in other frontend frameworks like "React" and Vue.js" and I was please to discover that Salesforce was planning to get more inline with current web standards with LWC.

### Etienne Rocheleau

> > At Coveo for 4 years, my job consists of the exploration of all new technology from Salesforce. I am always looking for new ways to integrate Coveo with Salesforce to create a better experience for users. I produce a lot of demos, POCs, and prototypes as part of my exploration and I really love when people find new and interesting ways to use Coveo to solve an issue they have. I am very familiar with Salesforce technologies (Aura, Apex, Visualforce, Lightning, etc.) and outside of that, I have experience with JavaScript/TypeScript and some web frameworks such as React, Angular and others.

Author: I'm trying okay... ;)

\- Etienne

> From a developer point of view, switching from Aura to LWC meant an access to better tools and following current web standards. We had put a lot of efforts to be able to have relatively good tests with Aura components, but with LWC it was part of the framework which in turn would mean a faster release cycle and higher product quality. To compare both, with Aura we could mainly do integration tests but we had to deploy the code to a Salesforce organization in order to test it. With LWC I knew we could do local unit tests which meant a lot quicker developement.
> 

> > Also because I was already used to other web frameworks, I liked the simplicity of the boilerplate needed for LWC. Compared to Aura from a readability standpoint, LWC is better because of how the code is written. It makes the code clearer and clearer code means less mistakes.

Reviewer: ```suggestion
> Also because I was already used to other web frameworks, I liked the reduced amount of boilerplate needed for LWC. Compared to Aura from a readability standpoint, LWC is better because of how the code is written. It makes the code clearer and clearer code means less mistakes.
```

> \- Jeremie

Reviewer: ```suggestion
\- Jérémie
```

## What did you decide to create as your first experience with LWC?

> > Our Salesforce integration uses our own UI framework we have developed in house called the [Coveo Search-UI](https://github.com/coveo/search-ui). It is a collection of UI components all written in pure TypeScript that are used to build a search page. Our first component was a wrapper around this framework to integrate it within Salesforce and with LWC. This component is responsible of loading the required Static Resources and then initializing our framework.

Reviewer: ```suggestion
> Our Salesforce integration uses our own open source UI framework we have developed in house called the [Coveo Search-UI](https://github.com/coveo/search-ui). It is a collection of UI components all written in pure TypeScript that are used to build a search page. Our first component was a wrapper around this framework to integrate it within Salesforce and with LWC. This component is responsible of loading the required Static Resources and then initializing our framework.
```

\- Jeremie

> ```js

Author: I tried something with it, check updated PR please.

    this.allScriptsLoaded = this.loadScripts();
  }

  loadStyles() {
    return loadStyle(this, `${SEARCH_UI}/css/CoveoFullSearchNewDesign.css`);
  }

  loadScripts() {
    // Prevent reloading of the library.
    if (window.Coveo) {
      return Promise.resolve();
    }
    return loadScript(this, `${SEARCH_UI}/js/CoveoJsSearch.js`))
      .then(() => loadScript(this, `${SEARCH_UI}/js/templates/templatesNew.js`));
  }

  renderedCallback() {
    Promise.all([this.allScriptsLoaded, this.stylesLoaded]).then(() => {
      // Everything is loaded, start the initialization of our framework.
      this.initializeSearchUI();
    });
  }
}
```

```html
<template>
  <lightning-card>
    <div class='search-ui-container' lwc:dom="manual"></div>
  </lightning-card>
</template>
```

> Above is an example of our first LWC component. It loads the CSS and JS files we need on the `connectedCallback` which is a lifecycle method that is called on component initialization by Salesforce. And then on the `renderedCallback` it initializes our framework, this method is also called by Salesforce when a component is rendered (our framework is designed to only initialize itself once). 
>

> > Unfortunately, we cannot rely on some of the DOM improvements provided by the shadowDOM since we handle our own rendering loop of our components. This is done in the template file by activating `lwc:dom="manual"` on our element container. But since we are already pure JavaScript and following web design principles, it is still really fast.

Author: Changed to 
> we handle ourselves the rendering loop of our components

\- Etienne

## What was your first impression?

> First impression was to say Wow, I can actually base my code off of the MDN standards since it respects the recent web standards. I was really impressed by this because it meant all the *staying up to date* I was putting time on would be very useful with LWC. This felt like a brand new project that I could easily get started on. Compared to other emerging frameworks, there was a ton of documentation, learning paths (on [Trailhead](https://trailhead.salesforce.com/)) and examples on GitHub to find what you are looking for. I started small by doing trails on Trailhead with the GitHub repositories already setup for LWC. And I really appreciated how well all of it was integrated with [`sfdx`](https://developer.salesforce.com/tools/sfdxcli) and [VSCode](https://code.visualstudio.com/). It made it very simple to develop locally, push to a *scratch org* and test.

\- Louis

> > From the original exploration of LWC, I found that it made a lot of sense to use the declarative approach from the LWC framework. It is very clear, simple and elegant the way you describe a component's *API*. 

Author: Does it matter if it is, clear + (simple and elegant) or clear + simple + elegant? ;)

```js
import { LightningElement, api } from 'lwc';
import myApexMethod from '@salesforce/apex/MyClassName.myMethod';

export default class MyComponent extends LightningElement {
  @api 
  publicProperty;

    
  @wire(myApexMethod, { myParam: 'myValue' })
  wireHandler({data, error}) {
    // do something.
  };
}
```
> You clearly see what is exposed publicly outside of your component. It makes it easy to use composition of components, it is similar to React, but uses the shadow DOM to process rendering and to encapsulate behavior and data. And the `@wire` decorator makes it very simple to understand clearly that you are calling Apex declaratively and letting Salesforce take care of caching the information.

> \- Jeremie

Reviewer: should I bother?

## Can you go in more details on what changed compared to the way of doing things with Aura?

> With our use of LWC we also have taken a good look at our project structure that had a strong impact on our build process and automated deployment of changes. Our project's source code is organized by logical entities that we could call a *module*. For example, our AgentPanel or StandaloneSearchbox each regrouped both Salesforce specific files such as their respective Aura component but also some TypeScript code, Sass files and even Gulp tasks. So everything related to one *module* was in one folder. This forced us over time to use more and more Gulp tasks to transpile/compile/move files around to create a folder structure that respects the Salesforce expected folder structure. This Salesforce folder structure is organized instead around types, all Aura components together, all Static Resources together, etc.
> 

![Coveo SFINT folder structure]({{ site.baseurl }}/images/2019-04-15-LWC/project_structure.png)

> > With the new advantages provided by `sfdx` but also the improved intellisense from the LWC framework, we now have reason enough to look at making the switch to the Salesforce folder structure instead. This new structure simplifies a lot our build process because we can remove a lot of the manual files handling with Gulp. I still feel however that more freedom with how the files are organized could go a long way to facilitate adoption of LWC by different teams because each team will have their own folder structure.

Author: No to your 2nd point, Gulp is used to handle files manually, but not causing it.

changed it to :
> This new structure simplifies a lot our build process because we can remove a lot of the manual files handling we were doing with Gulp.

>
> Before `sfdx` we used to leverage [JSforce](https://jsforce.github.io/) to do some of the deployment, many utils function we created over the years and a custom JavaScript wrapper around `sfdx` [sfdx-js](https://www.npmjs.com/package/sfdx-js) to handle our deployment. Using this method, with every change we wanted to test, we had to deploy over 300 files with the Metadata API, it took long enough to be able to stay up to date on Hacker News in between deploys. Now with the `sfdx` folder structure, only files that have changed are redeployed which takes seconds now instead of minutes. Multiply this by the number of times we deploy per day and per the number of members of the team, it is a huge time gain.

\- Louis

> > Quote from Andre about previous pain points around not using sfdx/Aura components.

Reviewer: is that a TODO?

Reviewer 3: Upvote

Author: Commit mixed up, I had removed it in the other branch I guess... thanks.

\- Louis

> Some gotchas to take into account that had me searching for a little while:
>
> 1. As of writing this post, LWC do not allow for custom decorators, only the ones provided by Salesforce and the LWC framework. JavaScript decorators are still only a proposition to ECMAScript, but seeing the decorators offered by Salesforce makes me want to be able to create my own. For example, I would like to create a `@log({level: 'info'})` decorator to mark a method to log to the console for tracking purposes. Or one of the frequently seen usecase with LWC, `@executeOnce` could be useful too.
> 2. Be wary of the difference between `this.querySelector`, `this.template.querySelector`, and `element.shadowRoot.querySelector`. The first one is used to search the DOM in a *slot*, the second one is used to search within the body of your LWC, and the third one is used within unit tests.

\- Jeremie

> Don't be afraid to make "utils" component. LWC allows you to easily import multiple component into one so make sure your code is clean and well seperated. It also makes thing easier to test and allows you to re-use  specific part of your application in other components.

\- Andre

> > It feels very natural to develop with LWC. It seems like the design decisions make sense and the integration with Salesforce's backend feels seamless. I am sure it will be easier to ramp up a new member of the team with LWC. One of the best lesson learned over the years is to *keep it simple* and LWC lets us do that. We can have clearly defined building blocks (components) that are designed to be re-used by our multiple integration points within Salesforce. Our package structure is simpler and well organized because one component regroups HTML, JavaScript and CSS. And LWC grants us enough freedom to keep most of the (Coveo Search-UI)[https://github.com/coveo/search-ui] simply as a Static Resource. And that way we can control the DOM of our search page manually while the LWC integration is in charge of the connection to Salesforce's data and API that we need to provide a great user experience.

Reviewer 3: One of the best lessons (missing s)

Add comma between Javascript and "and CSS"

You reversed the brackets for the Search UI link ( should be `[name](link)` )

Remove the "And" at the beginning of the last sentence

Author: Just for you :) Otherwise nobody would understand then sentence :)

\- Etienne

## What will the future look like for the Coveo for Salesforce integration team?

> > The Coveo for Salesforce integration package grew a lot over the past years. We've obviously added a lot of new features, but with the transition from Visualforce to Aura, and now the transition to LWC, our package is full of relics. We look forward to committing more resources to LWC and phasing out more and more Visualforce bit by bit. With LWC being closer to web standards than either Visualforce and Aura, we hope to be able to grow the team even faster by bringing in more web developers into the Salesforce ecosystem. Looking back at the past years and seeing what our integration package will be like in a short time once we achieve what we are planning makes us excited to see what Salesforce will become in the future.

Reviewer 3: than either Visualforce or Aura (not and)

\- Andre

## Would you like to know more?

> ### Read more on [Web Components](https://developer.mozilla.org/en-US/docs/Web/Web_Components) (MDN)

Reviewer 3: I don't think those need to be headings

> ### Read more on [Lightning Web Components](https://developer.salesforce.com/docs/component-library/documentation/lwc/lwc.get_started_introduction) (developer.salesforce)

Reviewer 3: developer.salesforce.com looks less weird

Author: Yea but it renders as an url if I add a .com...

Changed it to (Salesforce developers) which is their "title"
